[#3] 로그인 및  로그아웃

haileyKimDev opened the pull request:
## 개요
로그인 및 로그아웃 기능을 구현하고, 타임리프로 메인페이지, 마이페이지를 타임리프로 만들었습니다. 

## 작업사항
- 스프링  시큐리티로 로그인 

## 변경로직
- UserDetailsService인터페이스를 구현하는 UserSecurityService 를 생성
- login페이지 및 로그인 후 mypage로 이동 적용
- SpringSecurityConfig 생성
- 로그인 성공 케이스 테스트 작성

haileyKimDev committed:
feat : spring security 설정
-스프링 시큐리티 설정
-로그인

haileyKimDev committed:
feat add login page
Related : #3

haileyKimDev committed:
feat: 로그인 테스트코드
- 로그인 성공 테스트 코드추가
related: #3

haileyKimDev committed:
feat: 로그인 테스트 코드
- 로그인성공 케이스 테스트작성
related: #3

haileyKimDev committed:
feat: UserSecurityService 리팩토링
- UserSecurityService의 loadUserByUsername 리턴값 정정
- Login 페이지 로그인실패시  에러 메세지 노출 related: #3

f-lab-fi commented on src/main/java/com/smart/user/service/UserService.java:
```diff
@@ -1,11 +1,14 @@
 
 public interface UserService {
 
   void join(UserDto.JoinRequest request);
 
   UserDto.Response getUserByEmail(String email);
 
+  Optional<User> findByEmail(String username);
```
패키지를 어떻게 나누느냐에 따라 달라질 수 있겠지만, 통상적으로 Service가 표현계층(Controller)와 밀접한 관계를 맺는건 피해야합니다. Response 같은 이름은 HTTP와 밀접한 관련을 가지고 있죠. 가급적 표현 계층에 의존적이지 않은 이름, 의존적이지 않은 필드들을 사용해야합니다. 관련해서는 정화님께도 드린 이야기처럼, 다다음 멘토링 시간에 패키지 구조에 대해 이야기 드릴때 설명 드릴 예정입니다.

haileyKimDev replied:
@f-lab-fi  네
UserDto의 이너클래스 Response는 유저 정보를 담고있어서 UserInfo로 변경하였습니다.
아래 PR 확인부탁드립니다 ! 
[새로운 PR 링크](https://github.com/f-lab-edu/s-mart/pull/16)

f-lab-fi commented on src/test/java/com/smart/user/controller/LoginControllerTest.java:
```diff
@@ -0,0 +1,67 @@
+import org.springframework.test.web.servlet.ResultActions;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+
+@SpringBootTest
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
+public class LoginControllerTest {
```
실패하는 상황에 대한 테스트도 추가해보셨으면 좋겠습니다.
현재 NotFoundUserException 같은 예외들도 던져주고 있죠?
이런 예외가 잘 던져지는 것 역시 테스트 해보시죠. :)

haileyKimDev replied:
넵 ! 실패하는 상황 테스트를 추가했습니다.
- UserSecurityServiceTest에 로그인 실패 예외 처리 테스트 케이스
- LoginControllerTest에 로그인 실패 시 리다이렉트하는 테스트 케이스를 작성했습니다!

haileyKimDev committed:
feat : 로그인 실패테스트추가
- 로그인 실패 케이스 테스트 추가
- 로그인 실패시 로그인 페이지에 머무르도록 변경
Related: #3

haileyKimDev committed:
feat: 로그인 실패 예외처리 테스트 추가
- UserSecurityServiceTest추가
- 로그인 실패시 NotFoundUserException호출 테스트추가
Related :#3

f-lab-fi commented on src/main/java/com/smart/user/service/UserSecurityService.java:
```diff
@@ -12,13 +13,15 @@
+    com.smart.user.domain.User user = userDao.getUserByEmail(email);
+    if (user == null) {
+      throw new NotFoundUserException();
+    }
```
오잉 이건 왜 이렇게 바뀌었을까요? ㅠㅠ
혹시 [이 리뷰](https://github.com/f-lab-edu/s-mart/pull/10#discussion_r1148580931) 때문일까요?

haileyKimDev replied:
제가 만든 Optional<User> findByEmail(String username)가 정화님이 이전에 만드신 getUserByEmail과 기능이 같아서
기존 메서드를 활용하는 걸로 바꾸었습니다.
Optional<User> findByEmail로 다시 적용하는게 나을까요 ?

f-lab-fi replied:
@haileykim2014 

아하 그렇군요~
똑같은 역할을 메서드라면 그걸로 대체하여 사용해도 괜찮아보입니다.
다만, Optional에 대해서는 여기서는 충분히 적용해볼만한 가치가 있어보입니다.

> Optional findByEmail로 다시 적용하는게 나을까요 ?

getUserByEmail가 Optional로 바뀌면 어떨까요?
우리가 어떤 반환 값에 대해 Optional을 적용해볼지 고민해야하는지 찾아보시면 좋을 것 같습니다!

haileyKimDev replied:
@f-lab-fi 

아 널체크 및 반환을 동시에 해놓고 코드를 늘리고있었군요...
기존 로직 재사용하려고 분기처리보다는 Optional로 기존로직을 변경하는 것이 더 좋았을 거라는걸 깨달았습니다.
다시 Optional 로 변경하겠습니다 !

hoa0217 committed:
feat: "develop conflict 해결"
- 충돌 해결
- UserViewController 생성
- 인증페이지 생성

Related: #3
